File: algorithms/20. Valid Parentheses/solution.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        pairs = {'(': ')', '[': ']', '{': '}'}
        stack = []
        for c in s:
            if c in pairs:
                stack.append(pairs[c])
            elif not stack or stack.pop() != c:
                return False
        return not stack

File: algorithms/20. Valid Parentheses/solution.cc
class Solution {
public:
    bool isValid(string s) {
        unordered_map<char, char> pairs({{'(', ')'}, {'[', ']'}, {'{', '}'}});
        stack<char> left;
        for ( char c : s ) {
            if ( pairs.find(c) != pairs.end() ) {
                left.push(pairs[c]);
            } else if ( left.empty() || left.top() != c ) {
                return false;
            } else {
                left.pop();
            }
        }
        return left.empty();
    }
};
